### ISSUE TITLE
p:escape-markup not properly escaping special characters (e.g. Arabic characters)

### ISSUE THREAD

**Issue by davidfox-ap** [ndw/xmlcalabash1#220]
Escaping XML containing special characters--in particular, foreign alphabets--does not preserve these characters in the escaped version. A sample pipeline is below:

```
<p:declare-step xmlns:p="http://www.w3.org/ns/xproc"
    xmlns:c="http://www.w3.org/ns/xproc-step" version="1.0">
    <p:serialization port="result" encoding="utf-8"/>
    <p:input port="source">
        <p:inline>
            <row id="1">
                <field id="2">Edita Food Industries</field>
                <field id="3">ايديتا للصناعات الغذائية</field>
            </row>
        </p:inline>
    </p:input>
    <p:output port="result"/>
    <p:escape-markup/>
    <p:wrap-sequence wrapper="test"/>
    <p:unescape-markup/>
</p:declare-step>
```

Output with latest version of Calabash:

```
<test>
    <field id="2">Edita Food Industries</field>
    <field id="4">├ÿ┬º├Ö┼á├ÿ┬»├Ö┼á├ÿ┬¬├ÿ┬º ├ÖΓÇ₧├ÖΓÇ₧├ÿ┬╡├ÖΓÇá├ÿ┬º├ÿ
        ┬╣├ÿ┬º├ÿ┬¬ ├ÿ┬º├ÖΓÇ₧├ÿ┬║├ÿ┬░├ÿ┬º├ÿ┬ª├Ö┼á├ÿ┬⌐</field>  
</test>
```

Expected output:

```
<test>
                <field id="2">Edita Food Industries</field>
                <field id="3">ايديتا للصناعات الغذائية</field>
</test>
```

I am escaping and unescaping here, only to make it easier to read. The results are the same if you skip the final 'unescape-markup' step.

**Comment by ndw** [comment]
I can't reproduce this. What version of Java are you using, and what version of XML Calabash?

Is the pipeline encoded in UTF-8?

It's not immediately obvious to me how this is related to the escape and unescape markup steps. If you just run the data through p:identity, do you get the same results?

**Comment by davidfox-ap** [comment]
Thanks for your reply. My Java version:

java version "1.8.0_40"
Java(TM) SE Runtime Environment (build 1.8.0_40-b26)

And I'm using Calabash: 1.1.4-95

In my actual workflow I am calling an external XML, but for testing purposes I put the problem data inline. This testing xproc is encoded in UTF-8 and when I swap out the escape funtion for p:identity the characters look good.

I've created two versions of the test pipeline (one escaped and one using identity) with their outputs:

https://www.dropbox.com/sh/iqdutzi15m1196i/AACDMIwqokebDiyCsljqUNw7a?dl=0

I'm new to XProc and Calabash so it's possible I'm just doing something incorrectly. If it doesn't seem like a bug to you I can move this discussion elsewhere. Thanks!

**Comment by davidfox-ap** [comment]
Hi Norm, after we spoke earlier this week I thought to try the same files, with the latest calabash, on a Linux machine (rather than Windows, where I started). The files I shared above both worked fine here. So, I wonder if it is a java version issue, particular to Windows. My application will eventually run in Linux, so I'm not sure why I began by testing in Windows. :-)

Thank you for your help!

**Comment by ndw** [comment]
I can't conveniently test on Windows so I'm going to close this out. If it becomes important to you (or someone else reading this) later, please feel free to re-open it and I'll investigate.

**Comment by raducoravu** [comment]
:+1: @wendellpiez reported the same problem. When you run encoding tests on Linux you can specify in the Java command line a parameter like this:

```
    -Dfile.encoding=ASCII
```

to force set the default platform encoding to a very restrictive encoding.
Could this issue be reopened?

**Comment by raducoravu** [comment]
In this particular case the problem is in the method: com.xmlcalabash.library.EscapeMarkup.run() at line:

```
            String data = outstr.toString();
```

where you are converting a stream of bytes to a string without specifying an encoding. This means that the default platform encoding will be used.
In this case I would do something like this:

```
            StringWriter sw = new StringWriter();
            serializer.setOutputWriter(sw);
            S9apiUtils.serialize(runtime, child, serializer);
            String data = sw.toString();
```

so I would instruct the serializer to save directly to a character stream.

**Comment by raducoravu** [comment]
Other places where you have potential similar problems:

```
  com.xmlcalabash.extensions.Zip.storeJSON(FileToZip, XdmNode, OutputStream)
  com.xmlcalabash.library.Store.storeJSON(XdmNode, String, String, String)
  com.xmlcalabash.util.NodeToBytes.storeJSON(XdmNode, OutputStream)
```

As JSON is UTF-8 encoded you would need to explicitly give UTF-8 encoding when creating the print writer (which writes in chars) over the output stream (which writes in bytes):

```
new PrintWriter(new OutputStreamWriter(os, "UTF8"));
```

**Comment by ndw** [comment]
Thanks. I'll take a look. I'm hoping to do some work on XML Calabash next week.

**Comment by raducoravu** [comment]
Great, we are in no hurry. If you fix this we'll probably ship the fix with Oxygen 18.1 in a couple of months.

**Comment by wendellpiez** [comment]
This is so awesome thanks!

**Comment by raducoravu** [comment]
:+1:

**Comment by raducoravu** [comment]
The problem I reported in "com\xmlcalabash\library\EscapeMarkup" still persists in 1.1.12. Was the fix made after the kit was built?

**Comment by raducoravu** [comment]
@ndw could you give me an answer to the question I asked above? Because the problem with EscapeMarkup still persists...

**Comment by ndw** [comment]
Indeed. I overlooked that case. I thought I had a test...grrr. Will fix now and publish 1.1.15 asap.

**Comment by raducoravu** [comment]
Thanks @ndw, much appreciated.
